src/rest_framework_jwt/utils.py

```diff
@@ -105,17 +120,51 @@ def jwt_decode_token(token):
     options = {
         'verify_exp': api_settings.JWT_VERIFY_EXPIRATION,
     }
-    # get user from token, BEFORE verification, to get user secret key
-    unverified_payload = jwt.decode(token, None, False)
-    secret_key = jwt_get_secret_key(unverified_payload)
-    return jwt.decode(
-        token, api_settings.JWT_PUBLIC_KEY or secret_key,
-        api_settings.JWT_VERIFY, options=options,
-        leeway=api_settings.JWT_LEEWAY, audience=api_settings.JWT_AUDIENCE,
-        issuer=api_settings.JWT_ISSUER, algorithms=[
-            api_settings.JWT_ALGORITHM
-        ]
-    )
+
+    algos = api_settings.JWT_ALGORITHM
+    if not isinstance(algos, list):
+        algos = [algos]
+
+    hdr = jwt.get_unverified_header(token)
+    alg = hdr["alg"]
+    if alg not in algos:
+        raise jwt.exceptions.InvalidAlgorithmError
+
+    kid = hdr["kid"] if "kid" in hdr else None
+
+    keys = None
+    if alg.startswith("HS"):
+        unverified_payload = jwt.decode(token, None, False)
+        keys = jwt_get_secret_key(unverified_payload)
+    else:
+        keys = api_settings.JWT_PUBLIC_KEY
+
+    # if we got named keys and a key id is present, we insist to have
+    # that one, otherwise we try all
+    if isinstance(keys, dict):
+        if kid:
+            try:
+                keys = keys[kid]
+            except:
+                raise jwt.exceptions.InvalidKeyError
+        else:
+            keys = list(keys.values())
+
+    if not isinstance(keys, list):
+        keys = [keys]
+
+    ex = None
+    for key in keys:
+        try:
+            return jwt.decode(
+                token, key, api_settings.JWT_VERIFY, options=options,
+                leeway=api_settings.JWT_LEEWAY,
+                audience=api_settings.JWT_AUDIENCE,
+                issuer=api_settings.JWT_ISSUER, algorithms=[alg]
+            )
+        except (jwt.exceptions.InvalidSignatureError) as e:
+                ex = e
```

Comment:
Why not raise the exception here?

Reply:
To try multiple keys for the case of no key id